sdk/python/feast/feature_store.py

```diff
@@ -805,28 +825,61 @@ def get_online_features(
         for entity in entities:
             entity_name_to_join_key_map[entity.name] = entity.join_key
 
+        needed_request_data_features = self._get_needed_request_data_features(
+            grouped_odfv_refs
+        )
+
         join_key_rows = []
+        request_data_features: Dict[str, List[Any]] = {}
+        # Entity rows may be either entities or request data.
         for row in entity_rows:
             join_key_row = {}
             for entity_name, entity_value in row.items():
+                # Found request data
+                if entity_name in needed_request_data_features:
+                    if entity_name not in request_data_features:
+                        request_data_features[entity_name] = []
+                    request_data_features[entity_name].append(entity_value)
+                    continue
                 try:
                     join_key = entity_name_to_join_key_map[entity_name]
                 except KeyError:
                     raise EntityNotFoundException(entity_name, self.project)
                 join_key_row[join_key] = entity_value
                 if entityless_case:
                     join_key_row[DUMMY_ENTITY_ID] = DUMMY_ENTITY_VAL
-            join_key_rows.append(join_key_row)
+            if len(join_key_row) > 0:
+                # May be empty if this entity row was request data
+                join_key_rows.append(join_key_row)
+
+        if len(needed_request_data_features) != len(request_data_features.keys()):
+            raise RequestDataNotFoundInEntityRowsException(
+                feature_names=needed_request_data_features
+            )
 
         entity_row_proto_list = _infer_online_entity_rows(join_key_rows)
 
-        union_of_entity_keys = []
+        union_of_entity_keys: List[EntityKeyProto] = []
         result_rows: List[GetOnlineFeaturesResponse.FieldValues] = []
 
         for entity_row_proto in entity_row_proto_list:
+            # Create a list of entity keys to filter down for each feature view at lookup time.
             union_of_entity_keys.append(_entity_row_to_key(entity_row_proto))
+            # Also create entity values to append to the result
             result_rows.append(_entity_row_to_field_values(entity_row_proto))
 
+        # Add more feature values to the existing result rows for the request data features
+        for feature_name, feature_values in request_data_features.items():
+            for row_idx, feature_value in enumerate(feature_values):
+                result_row = result_rows[row_idx]
+                result_row.fields[feature_name].CopyFrom(
+                    python_value_to_proto_value(feature_value)
+                )
+                result_row.statuses[
+                    feature_name
+                ] = GetOnlineFeaturesResponse.FieldStatus.PRESENT
+
```

Comment:
we should also assert these status are set correctly in tests.

Reply:
We don't seem to assert them now and it also doesn't seem to be exposed publically in the API?